[#49956] Accidentaly granting access to Nextcloud project folders that are no members of the project

wielinde opened the pull request:
https://community.openproject.org/work_packages/49956

This PR attempts to fix a problem with the access control lists of project folders in Nextcloud. Apparently a memoized object (for performance optimization) containing the ACL settings for Nextcloud users that have an admin pendant in OpenProject got unexpectedly extended with other users. The other users were members of other projects that had their persmissions already set in a previous loop for a different project. So the bug made memoized object grow with each setting of ACLs per project folder. Non admin users falsely were treated as admins. And admins always have access to all project folders.

wielinde committed:
[#49956] Accidentaly granting access to Nextcloud project folders that are no members of the project

https://community.openproject.org/work_packages/49956

wielinde commented on modules/storages/app/services/storages/group_folder_properties_sync_service.rb:
```diff
@@ -212,7 +212,7 @@ def project_folder_permissions(project:)
     end
     tokens_query.each_with_object({
-                                    users: admins_project_folder_permissions,
+                                    users: admins_project_folder_permissions.clone,
```
@ba1ash `admins_project_folder_permissions` is a memoized object. We need to clone it first, if we want to modify the object as we do in line 219. Othwerwise we also modify the memoized object and that would be a bad side effect.

aaron-contreras reviewed (approved):
Discussed the bug description with Wieland. The thesis adds up 👍 Only thing I'd like to see is a spec that demonstrates the current unexpected behavior and green with the fix.

aaron-contreras committed:
Refactor: Re-implement `#project_folder_permissions`

Define `#project_folder_permissions` as the combination of
- `#admins_project_folder_permissions` and
- `#members_project_folder_permissions`.

By doing so, we get to keep `@admins_project_folder_permissions`
memoized and by merging both hashes together, we ensure that:

1. We don't modify the memoized object
2. We return a new hash, ensuring that state is kept clean.

aaron-contreras committed:
Add spec to demonstrate expected behavior

The spec setup now ensures that "Yoda" is not bundled into the
project_folder_permissions of subsequent project folders.

aaron-contreras commented:
Hi, @wielinde I've stayed a bit after hours and refactored the service to hopefully make the intention a lot clearer and remove any hackiness or hidden assumptions. Also, I made sure to update the current specs so that they would do two things:

1. Demonstrate the buggy behavior with the former implementation of `#project_folder_permissions` by raising errors of un-stubbed requests (because the request body would include unexpected users).
2. Be green only if we don't cascade permissions and bundle permissions down to subsequent project folders.

ba1ash committed:
Remove unnesessary memoization.

There is nothing heavy to cache like network or db requests.
So, let's construct new object for admins permissions and then modify it
by calling merge! with project members permissions.

ba1ash committed:
Change variable names.

- Rename `single_project_member` to be `single_project_user` because
  actually the object inside is a user and we have members as well in
  the database as a separate entity. The reason is to reduce possible
  confusion for a bit.
- Rename `user` to be `multiple_projects_user`. The name has more
  meaning about how the user is used and supposed to be used.
  And the name is coherent with `single_project_member` now.

wielinde commented:
@aaron-contreras Thank you for improving the code and adding specs. Thank you @ba1ash for reviewing and simplifying it.
